consensus/src/lib.rs

```diff
@@ -1,27 +1,1112 @@
+            return Ok(connected_genesis);
+        }
+
+        if let Some(best_block_id) = best_block_id {
+            // Chain trust is higher than the best block
+            let current_best_block_index = self
+                .db_tx
+                .get_block_index(&best_block_id)
+                .map_err(BlockError::from)?
+                .expect("Inconsistent DB");
+            if new_block_index.get_chain_trust() > current_best_block_index.get_chain_trust() {
+                self.reorganize(Some(best_block_id), &new_block_index)?;
+                return Ok(Some(new_block_index));
+            } else {
+                // TODO: we don't store block indexes for blocks we don't accept, because this is PoS
+                self.db_tx.set_block_index(&new_block_index)?;
+            }
+        }
+        Ok(None)
+    }
+
+    fn get_block_proof(&self, _block: &Block) -> u128 {
+        //TODO: We have to make correct one
+        10
+    }
+
+    fn add_to_block_index(
+        &mut self,
+        block: &Block,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<BlockIndex, BlockError> {
+        let prev_block_index = if best_block_id.is_none() && block.is_genesis(self.chain_config) {
+            // Genesis case
+            None
+        } else {
+            match block.get_prev_block_id() {
+                Some(prev_block) => {
+                    self.db_tx.get_block_index(&prev_block.into()).map_err(BlockError::from)?
+                }
+                None => return Err(BlockError::Orphan),
+            }
+        };
+        // Set the block height
+        let height = match &prev_block_index {
+            Some(prev_block_index) => prev_block_index.get_block_height().next_height(),
+            None => BlockHeight::zero(),
+        };
+        // Set Time Max
+        let time_max = match &prev_block_index {
+            Some(prev_block_index) => std::cmp::max(
+                prev_block_index.get_block_time_max(),
+                block.get_block_time(),
+            ),
+            None => block.get_block_time(),
+        };
+        // Set Chain Trust
+        let chain_trust = match &prev_block_index {
+            Some(prev_block_index) => prev_block_index.get_chain_trust(),
+            None => 0,
+        } + self.get_block_proof(block);
+        let block_index = BlockIndex::new(block, chain_trust, height, time_max);
+        Ok(block_index)
+    }
+
+    #[allow(dead_code)]
+    fn accept_block(
+        &mut self,
+        block: &Block,
+        best_block_id: &Option<Id<Block>>,
+    ) -> Result<BlockIndex, BlockError> {
+        let block_index = self.add_to_block_index(block, best_block_id)?;
+        self.check_block_index(&block_index)?;
+        self.db_tx.set_block_index(&block_index).map_err(BlockError::from)?;
+        self.db_tx.add_block(block).map_err(BlockError::from)?;
+        Ok(block_index)
+    }
+
+    fn exist_block(&self, id: &Id<Block>) -> Result<bool, BlockError> {
```

Comment:
`block_exists()`?

Reply:
Thank you @altonen! I've just pushed some fixes according to your suggestions from the previous PR. 

Fixed!